core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,235 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+
+case class FrameWithEffects(frame: Frame, effects: FrameEffects, dirtyPodIds: Set[PodId]) {
+  def applyEffects(newEffects: FrameEffects): FrameWithEffects = {
+    // TODO - we need to handle status snapshots and create a mechanism to signal that all cache should be recomputed
+    val newDirty = dirtyPodIds ++ newEffects.reverseStateEvents.iterator.collect {
+      case podEvent: PodStateEvent => podEvent.id
+    }
+    copy(frame = frame.applyStateEffects(effects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```

Comment:
```suggestion
    copy(frame = frame.applyStateEffects(newEffects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```
This above looks like a bug to me.

Reply:
Also, I believe we conflate two things in _Effects_:
1) things that change the Frame
2) side effects (like Mesos calls)
Could we call the 2) - `Actions` to make their purpose clearer?

Reply:
I could go with intents, perhaps